### PR TITLE
Feature/announcements

### PR DESCRIPTION
**Pull request by vvooycik**
## What part of the app is affected?
* [X] Frontend
* [ ] Backend
* [ ] Infrastructure

## Screenshots or Video Recording (if appropriate):

image
image
image
image

### CODE REVIEW

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    const response = await makeRequest<
+      AnnouncementUpdateResponseDto,
+      AnnouncementUpdateRequestDto
+    >(announcementApiUrl, 'PATCH', announcement);
+    if (!response.ok) {
+      showErrorSnackbar();
+    }
+    showSuccessSnackbar('Ogłoszenie zostało zapisane');
+    setIsEditAnnouncementModalOpen(false);
+    await mutate();
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]: Let's move mutation before showing success

```suggestion
    await mutate();
    showSuccessSnackbar('Ogłoszenie zostało zapisane');
    setIsEditAnnouncementModalOpen(false);
```

File: backend/src/seeds/seed.announcement.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    announcement.createdAt = new Date();
+    this.entities.push(announcement);
+    return announcement;
+  }
+}
+
+export const announcementFactory = new AnnouncementFactory();
+
+export const seedAnnouncements = () => {
+  const toBeCreatedAt = new Date('10/1/22');
```
Comment (bibixx) [inline backend/src/seeds/seed.announcement.ts]:
```suggestion
  const toBeCreatedAt = new Date();
```

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    );
+    setDeleteModalLoading(false);
+    if (!response.ok) {
+      showErrorSnackbar();
+      return;
+    }
+    setDeleteModalLoading(false);
+    closeDeleteModal();
+    await mutate();
+    showSuccessSnackbar('Ogłoszenie zostało pomyślnie usunięte');
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]:
```suggestion
    await mutate();
    setDeleteModalLoading(false);
    closeDeleteModal();
    showSuccessSnackbar('Ogłoszenie zostało pomyślnie usunięte');
```

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  const makeRequest = useMakeRequestWithAuth();
+  const {
+    setLoading: setDeleteModalLoading,
+    isLoading: isDeleteModalLoading,
+    isOpen: isDeleteModalOpen,
+    openModal: openDeleteModal,
+    closeModal: closeDeleteModal,
+  } = useActionModal();
+
+  const [announcementToBeDeleted, setAnnouncementToBeDeleted] =
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]: Let's change it to `announcementIdToBeDeleted`

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    useState<number>();
+  const [announcementToBeEdited, setAnnouncementToBeEdited] =
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]: Let's change it to `announcementIdToBeEdited`

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    const response = await makeRequest<AnnouncementDeleteResponseDto>(
+      announcementsApiUrl,
+      'DELETE',
+    );
+    setDeleteModalLoading(false);
+    if (!response.ok) {
+      showErrorSnackbar();
+      return;
+    }
+    setDeleteModalLoading(false);
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]: Why are we setting loading twice?

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        body,
+      },
+    };
+    const announcementApiUrl = '/api/announcements';
+    const response = await makeRequest<
+      AnnouncementCreateResponseDto,
+      AnnouncementCreateRequestDto
+    >(announcementApiUrl, 'POST', announcement);
+    if (!response.ok) {
+      showErrorSnackbar();
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]: Probably we should make a return after this. Right now if you have an error, you'll get the "Failed creation successfully" situation 😉

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+  return (
+    <Flex flexDirection="column" height="100%">
+      <Toolbar
+        sx={{
+          pl: { sm: 2 },
+          pr: { xs: 1, sm: 1 },
+          mb: '2rem',
+        }}
+      >
+        <Typography sx={{ flex: '1 1 30%' }} variant="h6" component="h1">
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]: Why do we need the `30%`?

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+      <Box style={{ overflow: 'auto' }}>
+        <Stack direction="column">
+          {announcements.map((announcement: DtoAnnouncement) => {
+            return (
+              <AnnouncementCard
+                announcement={announcement}
+                handleEdit={() => handleEdit(announcement)}
+                handleDelete={() => handleDelete(announcement.id)}
+              />
+            );
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]: Suggestion
1. Remove braces
2. No need for typing the argument of the function that is passed to the `map` function

```suggestion
          {announcements.map((announcement) => (
            <AnnouncementCard
              announcement={announcement}
              handleEdit={() => handleEdit(announcement)}
              handleDelete={() => handleDelete(announcement.id)} />
          ))}
```

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+              <Icon>filter_list</Icon>
+            </IconButton>
+          </Tooltip>
+        </Stack>
+      </Toolbar>
+      <Box style={{ overflow: 'auto' }}>
+        <Stack direction="column">
+          {announcements.map((announcement: DtoAnnouncement) => {
+            return (
+              <AnnouncementCard
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx]: Elements lack `key` prop

File: frontend/src/views/Announcements/AnnouncementCard/AnnouncementCard.tsx [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  };
+
+  const onEdit = () => {
+    setIsMenuOpen(false);
+    handleEdit();
+  };
+
+  return (
+    <Card
+      key={announcement.id}
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementCard/AnnouncementCard.tsx]: The `key` prop is unneccessary here

File: frontend/src/views/Announcements/AnnouncementCard/AnnouncementCard.tsx [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+      sx={{
+        minWidth: '60%',
+        maxWidth: '60%',
+        margin: 'auto',
+        mb: '3rem',
+        bgColor: '#fafafa',
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementCard/AnnouncementCard.tsx:57]: This doesn't seem to be valid CSS

File: frontend/src/views/Announcements/AnnouncementCard/AnnouncementCard.tsx [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+              <Icon fontSize="small" color="disabled">
+                more_vert
+              </Icon>
+            </IconButton>
+          )
+        }
+        title={<Typography variant="h6">{announcement.subject}</Typography>}
+        subheader={`${formatLong(new Date(announcement.createdAt))}, ${
+          announcement.createdBy.firstName
+        } ${announcement.createdBy.lastName}`}
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementCard/AnnouncementCard.tsx]: Simmilar to avatar — let's break this string into a separate variable for better readability

File: frontend/src/views/Announcements/AnnouncementCard/AnnouncementCard.tsx [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        id={announcement.id}
+      />
+      <CardContent>
+        <Typography variant="body2" sx={{ mx: 'auto', maxWidth: '95%' }}>
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementCard/AnnouncementCard.tsx]: Are we sure about the `95%`. If you want to add spacing, it'd be a better idea to just padding/margin with set value (preferably multiple of 8px)

File: frontend/src/views/Announcements/AnnouncementsList.tsx [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+import { useActionModal } from '../../hooks/useActionModal/useActionModal';
+import { AnnouncementCard } from './AnnouncementCard/AnnouncementCard';
+import { useAuth } from '../../hooks/useAuth/useAuth';
+import { ActionModal } from '../../components/ActionModal/ActionModal';
+import { useCommonSnackbars } from '../../hooks/useCommonSnackbars/useCommonSnackbars';
+import { useMakeRequestWithAuth } from '../../hooks/useMakeRequestWithAuth/useMakeRequestWithAuth';
+import { AnnouncementFormModal } from './AnnouncementFormModal/AnnouncementFormModal';
+import { AnnouncementFormSchema } from './AnnouncementFormModal/AnnouncementFormModal.schema';
+
+export const AnnouncementsList = () => {
```
Comment (bibixx) [inline frontend/src/views/Announcements/AnnouncementsList.tsx:26]: General comment to the whole component — why don't we break the business logic in this component into multiple hooks? Namely `useEditAnnouncement`, `useCreateAnnouncement`, and `useDeleteAnnouncement`.

**Review by bibixx — COMMENTED** [review]
Also a bug report: New lines are not respected after creating an announcement

https://user-images.githubusercontent.com/14880213/201546438-1572bbf4-6d0d-4819-9765-9c52c5f105b0.mp4

File: frontend/src/views/Announcements/AnnouncementsList.tsx (hunk 5 above)
Comment (vvooycik) [reply frontend/src/views/Announcements/AnnouncementsList.tsx]: It's not Id though - for edit I store whole object

File: frontend/src/views/Announcements/AnnouncementsList.tsx (hunk 8 above)
Comment (vvooycik) [reply frontend/src/views/Announcements/AnnouncementsList.tsx]: Don't exactly know where did it come from - I'll be changing this here and in instructors list to match the trainees list

File: frontend/src/views/Announcements/AnnouncementsList.tsx (hunk 15 above)
Comment (vvooycik) [reply frontend/src/views/Announcements/AnnouncementsList.tsx:34]: Cause I am nowhere near having react developer intuition. Will work on this suggestion though - update coming

**Review by bibixx — APPROVED** [review]

**Pull request merged**